### PR TITLE
Feature add multiarch docker images

### LINKED ISSUE #532: Bug: Eclair node fails to start on M1 Mac
Has anyone gotten an Eclair node to work reliably on an M1 mac? LND works fine with the new arm image, and c-lightning works for some reason even without the arm image, but Eclair is hit or miss. I noticed that I can sometimes get Eclair to work when I start up an Eclair node, it fails to start, and then I start it again. Although even when I get it to start, opening channels to other nodes fails.

### PR DESCRIPTION
**Pull request by NonsoAmadi10**
Closes #532 

### Description
- Adds docker support for arm64 support for eclair and c-lightning docker images
- Updates the readme to show how this images can be built using the `docker buildx` command

### CODE REVIEW

**Comment by NonsoAmadi10** [comment]
@jamaljsr I have my first PR raised for the issue #532 , I look forward to you reviewing and giving your feedback

**Comment by jamaljsr** [comment]
Hi @NonsoAmadi10 thank you for working on this. Great timing as I am working on Polar this morning.

The Core Lightning image build worked perfectly.

I attempted to build the Eclair image and ran into this error. I'm on an M1 Mac.

```sh
$ cd docker/eclair/
$ docker buildx build --platform linux/amd64,linux/arm64 --build-arg ECLAIR_VERSION=0.8.0 -t polarlightning/eclair:0.8.0 --push .
[+] Building 3.0s (8/8) FINISHED                                                                              
 => [internal] load build definition from Dockerfile                                                     0.0s
 => => transferring dockerfile: 5.28kB                                                                   0.0s
 => [internal] load .dockerignore                                                                        0.0s
 => => transferring context: 2B                                                                          0.0s
 => CANCELED [linux/arm64 internal] load metadata for docker.io/library/openjdk:11.0.16-jre-slim         3.0s
 => CANCELED [linux/amd64 internal] load metadata for docker.io/library/openjdk:11.0.16-jre-slim         3.0s
 => CANCELED [linux/amd64 internal] load metadata for docker.io/adoptopenjdk/openjdk11:jdk-11.0.3_7-alp  3.0s
 => ERROR [linux/arm64 internal] load metadata for docker.io/adoptopenjdk/openjdk11:jdk-11.0.3_7-alpine  3.0s
 => [auth] library/openjdk:pull token for registry-1.docker.io                                           0.0s
 => [auth] adoptopenjdk/openjdk11:pull token for registry-1.docker.io                                    0.0s
------
 > [linux/arm64 internal] load metadata for docker.io/adoptopenjdk/openjdk11:jdk-11.0.3_7-alpine:
------
Dockerfile:4
--------------------
   2 |     # Original Source: https://github.com/ACINQ/eclair/blob/v0.3.3/Dockerfile
   3 |     #########################
   4 | >>> FROM adoptopenjdk/openjdk11:jdk-11.0.3_7-alpine as BUILD
   5 |     
   6 |     # Setup maven, we don't use https://hub.docker.com/_/maven/ as it declare .m2 as volume, we loose all mvn cache
--------------------
ERROR: failed to solve: adoptopenjdk/openjdk11:jdk-11.0.3_7-alpine: no match for platform in manifest sha256:012ff3a1fd9c9237b14f7522307f8f64a19da0f3db37015b3456dafa69a29c0e: not found
```

**Comment by NonsoAmadi10** [comment]
Oh, I apologise for this. The dockerfile.arm64 file was the new eclair image that runs on mac m1 chips. Give me a minute please, let me resolve this error

**Comment by jamaljsr** [comment]
Oh I spoke to soon. The Core Lightning build failed after 40 mins with this error:

```sh
------
Dockerfile:135
--------------------
 134 |     
 135 | >>> RUN ./configure --prefix=/tmp/lightning_install --enable-static && \
 136 | >>>     make DEVELOPER=${DEVELOPER} && \
 137 | >>>     /root/.local/bin/poetry run make install
 138 |     
--------------------

ERROR: failed to solve: process "/bin/sh -c ./configure --prefix=/tmp/lightning_install --enable-static &&     make DEVELOPER=${DEVELOPER} &&     /root/.local/bin/poetry run make install" did not complete successfully: exit code: 2
```

The full log is huge so attaching it as a file instead of pasting. 
[cln-build.log](https://github.com/jamaljsr/polar/files/11129975/cln-build.log)

**Comment by NonsoAmadi10** [comment]
Screenshot 2023-04-03 at 16 22 15
@jamaljsr clightning builds well, the only issue is the push since I do not have access to the docker login for polar. As for eclair, adoptopenjdk has stopped support for that version of openjdk and encourages users to use eclipse-termiun, I have done this and tested, it builds on my mac, however i do not use an m1 yet but the buildx command did not fail on my end, I would appreciate if you'd test both for me and leave your feedback

**Comment by jamaljsr** [comment]
Thanks for the updates.

The Core Lightning build will take a while.

 I tried the Eclair build again and I'm getting the same error. It doesn't look like it's using the `Dockerfile.arm64` file when I run the following command:
```sh
$ docker buildx build  --platform linux/amd64,linux/arm64 --build-arg ECLAIR_VERSION=0.8.0 -t polarlightning/eclair:0.8.0  --push  .
```
Do I need to tweak the command to get it to use the `arm64` file?

**Comment by NonsoAmadi10** [comment]
Oh I removed the dockerarm64 image but if we are to introduce it, we will need to add a `-f Dockerfile.arm64` this indicates that we do not want the docker build to use the default dockerfile but a custom file

**Comment by NonsoAmadi10** [comment]
You can check it and if it works, I can bring it back on this branch

**Comment by jamaljsr** [comment]
The clightning image did build successfully and was pushed to Docker Hub. Unfortunately, the container fails to startup when using this new image. Here are the startup logs:
```
2023-04-03T18:33:26.132Z DEBUG   lightningd: testing /usr/local/libexec/c-lightning/lightning_closingd
2023-04-03T18:33:26.133Z DEBUG   lightningd: testing /usr/local/libexec/c-lightning/lightning_connectd
2023-04-03T18:33:26.134Z DEBUG   lightningd: testing /usr/local/libexec/c-lightning/lightning_gossipd
2023-04-03T18:33:26.136Z DEBUG   lightningd: testing /usr/local/libexec/c-lightning/lightning_hsmd
2023-04-03T18:33:26.137Z DEBUG   lightningd: testing /usr/local/libexec/c-lightning/lightning_onchaind
2023-04-03T18:33:26.138Z DEBUG   lightningd: testing /usr/local/libexec/c-lightning/lightning_openingd
2023-04-03T18:33:26.139Z DEBUG   hsmd: pid 47, msgfd 47
2023-04-03T18:33:26.161Z DEBUG   connectd: pid 48, msgfd 51
2023-04-03T18:33:26.161Z DEBUG   hsmd: Client: Received message 27 from client
2023-04-03T18:33:26.161Z DEBUG   hsmd: new_client: 0
2023-04-03T18:33:26.166Z DEBUG   connectd: Created listener on 172.22.0.3:19846
2023-04-03T18:33:26.166Z DEBUG   connectd: Created listener on 0.0.0.0:9735
2023-04-03T18:33:26.166Z DEBUG   connectd: REPLY WIRE_CONNECTD_INIT_REPLY with 0 fds
2023-04-03T18:33:26.166Z DEBUG   lightningd: io_break: connect_init_done
2023-04-03T18:33:26.166Z DEBUG   lightningd: io_loop: connectd_init
bitcoin-cli exited with code 255: qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
Make sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.
You can verify that your Bitcoin Core installation is ready for use by running:
2023-04-03T18:33:26.174Z **BROKEN** plugin-bcli: \nbitcoin-cli exited with code 255: qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory\n\n\nMake sure you have bitcoind running and that bitcoin-cli is able to connect to bitcoind.\n\nYou can verify that your Bitcoin Core installation is ready for use by running:\n\n    $ bitcoin-cli -regtest -rpcconnect=polar-n1-backend1 -rpcport=18443 -rpcuser=... -stdinrpcpass echo 'hello world'\n
    $ bitcoin-cli -regtest -rpcconnect=polar-n1-backend1 -rpcport=18443 -rpcuser=... -stdinrpcpass echo 'hello world'
2023-04-03T18:33:26.174Z INFO    plugin-bcli: Killing plugin: exited before we sent init
The Bitcoin backend died.
```

**Comment by jamaljsr** [comment]
For Eclair, there is still a `Dockerfile.arm64` file in this PR. If I try to use that one for the build, I get this error:
```sh
$ docker buildx build  --platform linux/amd64,linux/arm64 --build-arg ECLAIR_VERSION=0.8.0 -t polarlightning/eclair:0.8.0 -f Dockerfile.arm64 --push .
[+] Building 0.4s (29/51)                                                                                     
 => [internal] load .dockerignore                                                                        0.0s
 => => transferring context: 2B                                                                          0.0s
 => [internal] load build definition from Dockerfile.arm64                                               0.0s
 => => transferring dockerfile: 3.05kB                                                                   0.0s
 => [linux/arm64 internal] load metadata for docker.io/library/openjdk:11.0.4-jre-slim                   0.3s
 => [linux/arm64 internal] load metadata for docker.io/library/eclipse-temurin:11-jdk-jammy              0.2s
 => [linux/amd64 internal] load metadata for docker.io/library/eclipse-temurin:11-jdk-jammy              0.3s
 => [linux/amd64 internal] load metadata for docker.io/library/openjdk:11.0.4-jre-slim                   0.3s
 => [linux/amd64 stage-1 1/8] FROM docker.io/library/openjdk:11.0.4-jre-slim@sha256:1e4f3107324df951ae0  0.1s
 => => resolve docker.io/library/openjdk:11.0.4-jre-slim@sha256:1e4f3107324df951ae0adfd4f7018e831010411  0.0s
 => => sha256:b4024d25b93faceaf994f3b45bcc175de3b51dc56dc49a7555bbeb35fdca8782 0B / 42.07MB              0.0s
 => => sha256:e22147996cc08e68d0fa4d4ef9a441eaeef283ff4a2c64d1ef5e31340db246be 0B / 210B                 0.0s
 => => sha256:8d691f585fa8cec0eba196be460cfaffd69939782d6162986c3e0c5225d54f02 0B / 27.11MB              0.0s
 => => sha256:3da6fe7ff2ef0f58119eb2e8cb12c5c0b471527be95d43274b396f36fc5bbccd 0B / 3.25MB               0.0s
 => [internal] load build context                                                                        0.0s
 => => transferring context: 134B                                                                        0.0s
 => [linux/amd64 build  1/14] FROM docker.io/library/eclipse-temurin:11-jdk-jammy@sha256:80bee9bf721eb2  0.0s
 => => resolve docker.io/library/eclipse-temurin:11-jdk-jammy@sha256:80bee9bf721eb2bfb434019cc8fa5c4210  0.0s
 => [linux/arm64 stage-1 1/8] FROM docker.io/library/openjdk:11.0.4-jre-slim@sha256:1e4f3107324df951ae0  0.0s
 => => resolve docker.io/library/openjdk:11.0.4-jre-slim@sha256:1e4f3107324df951ae0adfd4f7018e831010411  0.1s
 => [linux/arm64 build  1/14] FROM docker.io/library/eclipse-temurin:11-jdk-jammy@sha256:80bee9bf721eb2  0.0s
 => => resolve docker.io/library/eclipse-temurin:11-jdk-jammy@sha256:80bee9bf721eb2bfb434019cc8fa5c4210  0.0s
 => CACHED [linux/arm64 build  2/14] RUN apt update && apt install -y curl tar bash                      0.0s
 => CACHED [linux/arm64 build  3/14] RUN mkdir -p /usr/share/maven /usr/share/maven/ref   && curl -fsSL  0.0s
 => CACHED [linux/arm64 build  4/14] WORKDIR /usr/src                                                    0.0s
 => ERROR [linux/arm64 build  5/14] COPY pom.xml pom.xml                                                 0.0s
 => CACHED [linux/amd64 build  2/14] RUN apt update && apt install -y curl tar bash                      0.0s
 => CACHED [linux/amd64 build  3/14] RUN mkdir -p /usr/share/maven /usr/share/maven/ref   && curl -fsSL  0.0s
 => CACHED [linux/amd64 build  4/14] WORKDIR /usr/src                                                    0.0s
 => ERROR [linux/amd64 build  5/14] COPY pom.xml pom.xml                                                 0.0s
 => ERROR [linux/amd64 build  6/14] COPY eclair-core/pom.xml eclair-core/pom.xml                         0.0s
 => ERROR [linux/amd64 build  7/14] COPY eclair-front/pom.xml eclair-front/pom.xml                       0.0s
 => ERROR [linux/arm64 build  6/14] COPY eclair-core/pom.xml eclair-core/pom.xml                         0.0s
 => ERROR [linux/arm64 build  7/14] COPY eclair-front/pom.xml eclair-front/pom.xml                       0.0s
 => ERROR [linux/arm64 build  8/14] COPY eclair-node/pom.xml eclair-node/pom.xml                         0.0s
 => ERROR [linux/arm64 build  9/14] COPY eclair-node/modules/assembly.xml eclair-node/modules/assembly.  0.0s
 => ERROR [linux/amd64 build  8/14] COPY eclair-node/pom.xml eclair-node/pom.xml                         0.0s
 => ERROR [linux/amd64 build  9/14] COPY eclair-node/modules/assembly.xml eclair-node/modules/assembly.  0.0s
 => CANCELED [linux/arm64 stage-1 2/8] WORKDIR /app                                                      0.0s
 => CANCELED [linux/amd64 stage-1 2/8] WORKDIR /app                                                      0.0s
------
 > [linux/arm64 build  5/14] COPY pom.xml pom.xml:
------
------
 > [linux/amd64 build  5/14] COPY pom.xml pom.xml:
------
------
 > [linux/amd64 build  6/14] COPY eclair-core/pom.xml eclair-core/pom.xml:
------
------
 > [linux/amd64 build  7/14] COPY eclair-front/pom.xml eclair-front/pom.xml:
------
------
 > [linux/arm64 build  6/14] COPY eclair-core/pom.xml eclair-core/pom.xml:
------
------
 > [linux/arm64 build  7/14] COPY eclair-front/pom.xml eclair-front/pom.xml:
------
------
 > [linux/arm64 build  8/14] COPY eclair-node/pom.xml eclair-node/pom.xml:
------
------
 > [linux/arm64 build  9/14] COPY eclair-node/modules/assembly.xml eclair-node/modules/assembly.xml:
------
------
 > [linux/amd64 build  8/14] COPY eclair-node/pom.xml eclair-node/pom.xml:
------
------
 > [linux/amd64 build  9/14] COPY eclair-node/modules/assembly.xml eclair-node/modules/assembly.xml:
------
Dockerfile.arm64:31
--------------------
  29 |     COPY eclair-front/pom.xml eclair-front/pom.xml
  30 |     COPY eclair-node/pom.xml eclair-node/pom.xml
  31 | >>> COPY eclair-node/modules/assembly.xml eclair-node/modules/assembly.xml
  32 |     RUN mkdir -p eclair-core/src/main/scala && touch eclair-core/src/main/scala/empty.scala
  33 |     # Blank build. We only care about eclair-node, and we use install because eclair-node depends on eclair-core
--------------------
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref zjo8dicalmye14cwlnzdu4248::yczoag9olwkdoyxhim5bdic86: "/eclair-node/modules/assembly.xml": not found
```

**Comment by NonsoAmadi10** [comment]
Oh it is a small bug. Let me check it and update. No luck with core lightning still??

**Comment by jamaljsr** [comment]
I posted my current issue with clightning in https://github.com/jamaljsr/polar/pull/701#issuecomment-1494797055

**Comment by NonsoAmadi10** [comment]
I have seen the core lightning issue. It shows it requires bitcoind, I would look into that. It shows progress though. Let me look at the eclair now

**Comment by NonsoAmadi10** [comment]
I just made a small change especiall for clightning, i added the arm64 image based on how it was defined on their repo with the modification polar likes to make. For eclair, the default dockerfile should work now for both machines

**Comment by jamaljsr** [comment]
Are you able to build these images on your machine? 

On my M1 Mac, the clightning build succeeded in a little over an hour, but it's still facing that `qemu-x86_64` error when trying to start it. On my Intel Mac, the build build has been running for over 3 hours now and it's still not done. 

The Eclair build has been running for over an hour on my M1 Mac and appears to be frozen. I may have to restart Docker or something. I haven't gotten a chance to test it on my Intel Mac yet.

**Comment by jamaljsr** [comment]
The clightning build on my Intel Mac finally finished and the `amd64` image does work on that machine. However, when I pull the `arm64` image on my M1 Mac, I am still facing the same `qemu-x86_64` error. 

I was finally able to build the eclair image using my Intel Mac. It took over an our to complete but it works on both Intel & M1 Macs 🙌.

**Comment by NonsoAmadi10** [comment]
This is goodnews!! What is left now is just c-lightning. I will tackle this today, we are almost there

**Comment by NonsoAmadi10** [comment]
I use a mac 2017 and here in africa internet is a problem. so what i do is I test on a vps because of the bandwidth there but i don't get your errors

**Comment by jamaljsr** [comment]
I have good news! I have successfully built a multi-arch image for Core Lightning using the `docker manifest` approach. You can see the image on Docker Hub. https://hub.docker.com/r/polarlightning/clightning/tags

In order to speed up the build process, I built the `arm64` image on my M1 machine and built the `amd64` image on my Intel machine. These are the commands I used with the two Dockerfiles in this PR.
```sh
# build amd64 image
$ docker build --build-arg CLN_VERSION=23.02.2 -t polarlightning/clightning:23.02.2-amd64 .
$ docker push polarlightning/clightning:23.02.2-amd64

# build arm64 image
$ docker build --build-arg CLN_VERSION=23.02.2 -t polarlightning/clightning:23.02.2-arm64 -f Dockerfile.arm64 .
$ docker push polarlightning/clightning:23.02.2-arm64

# combine into a single multi-arch image
$ docker manifest create polarlightning/clightning:23.02.2 --amend polarlightning/clightning:23.02.2-arm64 --amend polarlightning/clightning:23.02.2-amd64
$ docker manifest push polarlightning/clightning:23.02.2
```

**Comment by NonsoAmadi10** [comment]
Wow this is great!! Was there anything you added to the docker file for this to work?

File: docker/clightning/Dockerfile.arm64
Comment (jamaljsr) [inline docker/clightning/Dockerfile.arm64:1]: Would you mind adding the "Polar Modification" comments to this file, similar to the [Dockerfile](https://github.com/jamaljsr/polar/blob/master/docker/clightning/Dockerfile#L113-L121), to make it clear where the original source of this file is from and which parts were modified from the original? This makes it much easier to update it in the future when the source `Dockerfile` changes.

File: docker/clightning/Dockerfile.arm64 [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        ca-certificates \
+        curl \
+        dirmngr \
+        gettext \
+        git \
+        gnupg \
+        libpq-dev \
+        libtool \
+        libffi-dev \
+        python3 \
```
Comment (jamaljsr) [inline docker/clightning/Dockerfile.arm64:67]: I needed to add this on the line above `python3` in order for the build to succeed.
```
        protobuf-compiler \
```

In the `lightningd` repo, this line was added to the main `Dockerfile` in https://github.com/ElementsProject/lightning/pull/5917. It looks like they forgot to also update the other Dockerfiles at that time, so the one in their repo is broken.

File: docker/README.md [hunk 2]
```diff
@@ -23,17 +36,11 @@ _Warning: These images are not hardened and shouldn't be used to store real bitc
 
 ```sh
 $ cd bitcoind
-$ docker build --build-arg BITCOIN_VERSION=<version> -t polarlightning/bitcoind:<version> .
+$ docker buildx build  --platform linux/amd64,linux/arm64 --build-arg BITCOIN_VERSION=<version>  -t polarlightning/bitcoind:<version>  --push  .
```
Comment (jamaljsr) [inline docker/README.md:39]: There are some double spaces in this command. Can you remove them?

File: docker/README.md [hunk 3]
```diff
@@ -68,7 +75,7 @@ $ docker push polarlightning/bitcoind:<version>
 
 ```sh
 $ cd lnd
-$ docker build --build-arg LND_VERSION=<version> -t polarlightning/lnd:<version> .
+$ docker buildx build  --platform linux/amd64,linux/arm64 --build-arg LND_VERSION=<version> -t polarlightning/lnd:<version>  --push  .
```
Comment (jamaljsr) [inline docker/README.md:78]: Double spaces here as well.

File: docker/README.md [hunk 4]
```diff
@@ -102,7 +109,7 @@ $ docker push polarlightning/lnd:<version>
 
 ```sh
 $ cd clightning
-$ docker build --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version> .
+$ docker buildx build --platform linux/amd64,linux/arm64 --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version> .
```
Comment (jamaljsr) [inline docker/README.md:112]: Change this command to:
```sh
# build amd64 image (note: this takes a long time on ARM machines)
$ docker build --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version>-amd64 .
$ docker push polarlightning/clightning:<version>-amd64

# build arm64 image (note: this takes a long time on x86/x64 machines)
$ docker build --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version>-arm64 -f Dockerfile.arm64 .
$ docker push polarlightning/clightning:<version>-arm64

# combine into a single multi-arch image
$ docker manifest create polarlightning/clightning:<version> --amend polarlightning/clightning:<version>-arm64 --amend polarlightning/clightning:<version>-amd64
$ docker manifest push polarlightning/clightning:<version>
```

File: docker/README.md [hunk 5]
```diff
@@ -131,6 +138,7 @@ $ docker push polarlightning/clightning:<version>
 ```sh
 $ cd eclair
 $ docker build --build-arg ECLAIR_VERSION=<version> -t polarlightning/eclair:<version> .
+$ docker buildx build  --platform linux/amd64,linux/arm64 --build-arg ECLAIR_VERSION=0.6.2 -t polarlightning/eclair:0.6.2  --push  .
```
Comment (jamaljsr) [inline docker/README.md:141]: Remove the `docker build` line above this one and change `0.6.2` to `<version>`.

File: docker/README.md [hunk 6]
```diff
@@ -163,4 +171,4 @@ Once the updated `nodes.json` file is committed to master, the new images can be
 1. In the Network Designer sidebar, click on the **Show All Versions** toggle
 1. At the bottom of the node list, click on the **Check for new Node Versions** link
 1. A dialog will open displaying the new versions available
-1. Click the **Add New Versions** button to begin using them
+1. Click the **Add New Versions** button to begin using them
```
Comment (jamaljsr) [inline docker/README.md:174]: I think this just removed the last newline char from this line. Can you add it back to remove this diff?

**Review by jamaljsr — CHANGES_REQUESTED** [review]
Now that we've got all of the builds working, I've done a file-by-file review of your changes.

Can you also rebase this branch on `master` as I added LND `v0.16.0-beta` since you forked.

> Wow this is great!! Was there anything you added to the docker file for this to work?

There was only one small change I needed to make. I noted it below.

File: docker/README.md (hunk 5 above)
Comment (NonsoAmadi10) [reply docker/README.md]: I have done this, please review

File: docker/README.md (hunk 4 above)
Comment (NonsoAmadi10) [reply docker/README.md]: I have added this too, please review

File: docker/README.md (hunk 3 above)
Comment (NonsoAmadi10) [reply docker/README.md]: Removed the double spaces as well

File: docker/README.md (hunk 2 above)
Comment (NonsoAmadi10) [reply docker/README.md]: removed the spaces, please review

File: docker/clightning/Dockerfile.arm64
Comment (NonsoAmadi10) [reply docker/clightning/Dockerfile.arm64:1]: Sure!

File: docker/clightning/Dockerfile.arm64 (hunk 1 above)
Comment (jamaljsr) [reply docker/clightning/Dockerfile.arm64:67]: To be consistent with the [linuxarm64v8.Dockerfile](https://github.com/ElementsProject/lightning/blob/master/contrib/docker/linuxarm64v8.Dockerfile), can you move this one line up above `python3`.

File: docker/README.md (hunk 5 above)
Comment (jamaljsr) [reply docker/README.md]: Confirmed 👍

**Review by jamaljsr — COMMENTED** [review]
Ok, I have another round of minor updates just to cleanup the files. I think this should be good to go after these changes.

Also, can you squash your commits into one to make the history cleaner when this gets merged. Using `git rebase -i master` is the simplest method.

**Comment by NonsoAmadi10** [comment]
I have been able to squash, the only issue was that it required a force push😳

**Comment by jamaljsr** [comment]
The problem you are facing is because you merged in commits from `master` into this branch. Those need to be dropped instead of squashed. When you run `git rebase -i master`, set the rebase commands to the following:

```
pick 4ffbdf1d5 [feature]: add support for eclair and c-lightning for arm64 machines
squash 4adf378b7 chore(remove-me): tmp commit to publish to github
squash a8084bf42 chore(replace-base-images): replace base images for eclair and c-lightning
squash fb6e35c6f chore(replace-base-images): replace base images for eclair and c-lightning
squash 6c0f60763 chore: change base images
squash 56cc239ec chore(replace-base-images): replace base images for eclair and c-lightning
drop d6897bae4 chore(deps): update dependency stylelint-config-standard to v32
drop 35d9d0cd6 feat(images): add support for LND v0.16.0-beta
drop 6445103e0 chore(deps): update dependency i18next to v22.4.14
drop 951b00d4c chore(deps): update typescript-eslint monorepo to v5.57.1
drop aaba314be chore(deps): update dependency @types/react to v17.0.56
squash f99344d12 chore: modify dockerfile for clightning
```

Then you'll need to merge all of the commit messages into one. This would be a good message for this PR:
```
feat(docker): update Dockerfiles to build multi-arch images 
```

Finally, you'll need to force-push, which is perfectly fine in PRs. When that is done, you should only see one commit in the Commits tab of this PR.

**Comment by jamaljsr** [comment]
> Deleted the prior comment because I realized my local `master` branch was outdated, so I was likely seeing something different than you.

Github still shows 8 commits in this PR branch. You'll want to squash those down to one. First, make sure that your local `master` branch is pointing to the latest commit on Github https://github.com/jamaljsr/polar/commit/aaba314be07bf30df370d6ad53ef756f66a85238. Then when you run `git rebase -i master`, you should only see your 8 commits. You just need to squash all but the first one using the following rebase commands.

```
pick 4ffbdf1d5 [feature]: add support for eclair and c-lightning for arm64 machines
squash 4adf378b7 chore(remove-me): tmp commit to publish to github
squash a8084bf42 chore(replace-base-images): replace base images for eclair and c-lightning
squash fb6e35c6f chore(replace-base-images): replace base images for eclair and c-lightning
squash 6c0f60763 chore: change base images
squash 56cc239ec chore(replace-base-images): replace base images for eclair and c-lightning
squash f99344d12 chore: modify dockerfile for clightning
```

After saving and closing that document, then you'll need to merge all of the commit messages into one. This would be a good message for this PR:

```
feat(docker): update Dockerfiles to build multi-arch images 
```

Finally, you'll need to force-push, which is perfectly fine in PRs. When that is done, you should only see one commit in the Commits tab of this PR.

File: docker/README.md [hunk 7]
```diff
@@ -100,19 +103,33 @@ $ docker push polarlightning/lnd:<version>
...
+$ docker push polarlightning/clightning:<version>-amd64
+
+# build arm64 image (note: this takes a long time on x86/x64 machines)
+$ docker build --platform linx/arm64 --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version>-arm64 -f Dockerfile.arm64 .
+>>>>>>> 50964e37 (chore: modify dockerfile for clightning)
+$ docker push polarlightning/clightning:<version>-arm64
+
+# combine into a single multi-arch image
+$ docker manifest create polarlightning/clightning:<version> --amend polarlightning/clightning:<version>-arm64 --amend polarlightning/clightning:<version>-amd64
+$ docker manifest push polarlightning/clightning:<version>
```
Comment (jamaljsr) [inline docker/README.md:128]: You left a merge conflict in here that should be resolved.

**Comment by NonsoAmadi10** [comment]
I have done this @jamaljsr

File: docker/clightning/Dockerfile.arm64 [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+# BEGIN ElementsProject/lightning/contrib/docker/linuxarm64v8.Dockerfile
+#########################
+
+# This dockerfile is meant to compile a core-lightning x64 image
+# It is using multi stage build:
+# * downloader: Download litecoin/bitcoin and qemu binaries needed for core-lightning
+# * builder: Compile core-lightning dependencies, then core-lightning itself with static linking
+# * final: Copy the binaries required at runtime
+# The resulting image uploaded to dockerhub will only contain what is needed for runtime.
+# From the root of the repository, run "docker build -t yourimage:yourtag -f contrib/linuxarm64v8.Dockerfile ."
```
Comment (jamaljsr) [inline docker/clightning/Dockerfile.arm64:18]: Hate to be nitpicky, but there's still a diff in these comments. Can you paste in these lines which I copied from the [linuxarm64v8.Dockerfile](https://github.com/ElementsProject/lightning/blob/v23.02.2/contrib/docker/linuxarm64v8.Dockerfile) file.

```
# This dockerfile is meant to cross compile with a x64 machine for a arm64v8 host
# It is using multi stage build: 
# * downloader: Download litecoin/bitcoin and qemu binaries needed for Core Lightning
# * builder: Cross compile Core Lightning dependencies, then Core Lightning itself with static linking
# * final: Copy the binaries required at runtime
# The resulting image uploaded to dockerhub will only contain what is needed for runtime.
# From the root of the repository, run "docker build -t yourimage:yourtag -f contrib/linuxarm64v8.Dockerfile ."
```

File: docker/README.md [hunk 9]
```diff
@@ -68,17 +75,13 @@ $ docker push polarlightning/bitcoind:<version>
 
 ```sh
 $ cd lnd
-$ docker build --build-arg LND_VERSION=<version> -t polarlightning/lnd:<version> .
+$ docker buildx build --platform linux/amd64,linux/arm64 --build-arg LND_VERSION=<version> -t polarlightning/lnd:<version> --push  .
 ```
 
 Replace `<version>` with the desired LND version (ex: `0.7.1-beta`)
 
 **Push to Docker Hub**
```
Comment (jamaljsr) [inline docker/README.md:83]: Forgot to remove this line.

**Review by jamaljsr — COMMENTED** [review]
Nice job on the squash. The commit looks great!

Just have a few more minor changes and this should be ready to merge.

**Comment by NonsoAmadi10** [comment]
I have made the changes you requested. Please review again

**Comment by jamaljsr** [comment]
Oh no. You're last [push](https://github.com/jamaljsr/polar/compare/1c06b66ee7c8264917601b6ef349bee742a198a9..1ae649302f1ac542488f7059000d8ed20fdfef2a) just made a ton of changes to the `Dockerfile.arm64` file. What happened? Can you revert all of those changes?

You can compare your file to the original on a site like https://www.diffchecker.com. Compare the source [linuxarm64v8.Dockerfile](https://github.com/ElementsProject/lightning/blob/master/contrib/docker/linuxarm64v8.Dockerfile) file to your `Dockerfile.arm64` and the only changes you should see are the Polar modifications.

File: docker/README.md [hunk 10]
```diff
@@ -100,19 +101,25 @@ $ docker push polarlightning/lnd:<version>
...
-```
+# build amd64 image (note: this takes a long time on ARM machines)
+$ docker build --platform linux/amd64 --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version>-amd64 .
+$ docker push polarlightning/clightning:<version>-amd64
 
-Replace `<version>` with the desired c-lightning version (ex: `0.8.0`).
-
-**Push to Docker Hub**
+# build arm64 image (note: this takes a long time on x86/x64 machines)
+$ docker build --platform linx/arm64 --build-arg CLN_VERSION=<version> -t polarlightning/clightning:<version>-arm64 -f Dockerfile.arm64 .
```
Comment (jamaljsr) [inline docker/README.md:113]: I am doing one final round of builds for all nodes and discovered this small typo. Can you change `linx/arm64` to `linux/arm64`?

**Comment by NonsoAmadi10** [comment]
I have done this @jamaljsr

**Review by jamaljsr — APPROVED** [review]
tACK This is good to go 🚀

Thank you so much for spending all of this time iterating on this PR. I've been wanting to make these updates for a long while.

**Comment by NonsoAmadi10** [comment]
Thank you Jamal for giving me the opportunity. I would be looking to pick other issues with confidence now

**Pull request merged**